File: countdown/numbers.py
"""
Countdown (Cijfers&Letters)
Number game
https://www.datagenetics.com/blog/august32014/index.html
"""
import random

from countdown.utils import remove_random_element


def generate(big_count):
    """ Generate the Numbers problem
    return a tuple where the first element is the target and the second a list
    of numbers to make that target.
    """
    big_numbers = [25, 50, 75, 100]
    small_numbers = [1 , 1 , 2 , 2 , 3 , 3 , 4 , 4 , 5 , 5 , 6 , 6 , 7 , 7 , 8 , 8 , 9 , 9 , 10 , 10]
    total_amount_of_numbers = 6
    small_count = total_amount_of_numbers - big_count
    if big_count + small_count != total_amount_of_numbers:
        raise Exception(f"total amount of numbers must be {total_amount_of_numbers}")
    selection = []
    for i in range(0, big_count):
        selection.append(remove_random_element(big_numbers))
    for i in range(0, small_count):
        selection.append(remove_random_element(small_numbers))
    target = random.choice(range(101, 1000))
    return (target, selection)

File: internal/numbers/numbers.go
package numbers

import (
	"math/rand"
	"slices"
)

type TplVarsNumbers struct {
	Game      string
	Target    []int
	Selection []int
}

var bigNumbers = []int{25, 50, 75, 100}
var smallNumbers = []int{1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8, 8, 9, 9, 10, 10}
var totalNumberCount = 6

func Pick(bigCount int) ([]int, []int) {
	smallCount := totalNumberCount - bigCount
	target := -1
	selection := []int{}
	big := slices.Clone(bigNumbers)
	small := slices.Clone(smallNumbers)

	for range bigCount {
		iRnd := rand.Intn(len(big))
		selection = append(selection, big[iRnd])
		big = slices.Delete(big, iRnd, iRnd+1)
	}

	for range smallCount {
		iRnd := rand.Intn(len(small))
		selection = append(selection, small[iRnd])
		small = slices.Delete(small, iRnd, iRnd+1)
	}

	for target <= 100 {
		target = rand.Intn(1000)
	}

	return splitInt(target), selection
}

func splitInt(n int) []int {
	if n == 0 {
		return []int{0}
	}
	tmp := n
	length := 0
	for tmp > 0 {
		tmp /= 10
		length++
	}
	digits := make([]int, length)
	for i := length - 1; i >= 0; i-- {
		digits[i] = n % 10
		n /= 10
	}
	return digits
}
